src/mlpack/methods/decision_tree/best_binary_numeric_split.hpp

```diff
@@ -63,13 +72,83 @@ class BestBinaryNumericSplit
       const WeightVecType& weights,
       const size_t minimumLeafSize,
       const double minimumGainSplit,
-      arma::vec& classProbabilities,
+      arma::vec& splitInfo,
+      AuxiliarySplitInfo& aux);
+
+  /**
+   * Check if we can split a node.  If we can split a node in a way that
+   * improves on 'bestGain', then we return the improved gain.  Otherwise we
+   * return the value 'bestGain'.  If a split is made, then splitInfo and aux
+   * may be modified.
+   *
+   * This overload is used only for regression tasks.
+   *
+   * @param bestGain Best gain seen so far (we'll only split if we find gain
+   *      better than this).
+   * @param data The dimension of data points to check for a split in.
+   * @param responses Responses for each point.
+   * @param weights Weights associated with responses.
+   * @param minimumLeafSize Minimum number of points in a leaf node for
+   *      splitting.
+   * @param minimumGainSplit Minimum gain split.
+   * @param splitInfo Stores split information on a successful split.
+   * @param aux Auxiliary split information, which may be modified on a
+   *      successful split.
+   */
+  template<bool UseWeights, typename VecType, typename ResponsesType,
+           typename WeightVecType>
+  static double SplitIfBetter(
+      const double bestGain,
+      const VecType& data,
+      const ResponsesType& responses,
+      const WeightVecType& weights,
+      const size_t minimumLeafSize,
+      const double minimumGainSplit,
+      double& splitInfo,
       AuxiliarySplitInfo& aux);
 
+  /**
+  * Check if we can split a node.  If we can split a node in a way that
+  * improves on 'bestGain', then we return the improved gain.  Otherwise we
+  * return the value 'bestGain'.  If a split is made, then splitInfo and aux
+  * may be modified.
+  *
+  * This overload is specialized only for MSEGain fitness function.
+  *
+  * @param bestGain Best gain seen so far (we'll only split if we find gain
+  *      better than this).
+  * @param data The dimension of data points to check for a split in.
+  * @param responses Responses for each point.
+  * @param weights Weights associated with responses.
+  * @param minimumLeafSize Minimum number of points in a leaf node for
+  *      splitting.
+  * @param minimumGainSplit Minimum gain split.
+  * @param splitInfo Stores split information on a successful split.
+  * @param aux Auxiliary split information, which may be modified on a
+  *      successful split.
+  */
+  template<bool UseWeights, typename VecType, typename ResponsesType,
+          typename WeightVecType>
+  typename std::enable_if<
```

Comment:
This one isn't marked `static`, but I think it should be.  However, when you do this, I expect you will have an ambiguous function call compilation error, because `MSEGain` can match both this overload and the other one.  Thus, you need to use `std::enable_if<>` with the other overload, with the negated conditional: `std::enable_if<!HasBinaryScanInitialize ... || !HasBinaryStep ..., double`.

Note also you can use SFINAE as default-valued arguments to the function, like this:

```
  SplitIfBetter(
      const double bestGain,
      const VecType& data,
      const ResponsesType& responses,
      const WeightVecType& weights,
      const size_t minimumLeafSize,
      const double minimumGainSplit,
      double& splitInfo,
      AuxiliarySplitInfo& /* aux */,
      typename std::enable_if<..., void>::type* = 0);
```

But, to my knowledge they both work the same way, so it doesn't make a difference which you want to use.

Reply:
Hey @rcurtin, I added the `static` keyword to the function and it did compile. But, when I added the negated condition to the regular overload, It is giving me an error that "no matching overload of SplitIfBetter could be found". Can you please take a look at what is going wrong here?